#TASK 1:  KNOWLEDGE DESIGN


# **IMPORTING ESSENTIAL LIBRARIES**



In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import spacy
nlp = spacy.load("en_core_web_sm")
import re
from collections import Counter
from sklearn.metrics import precision_score, recall_score
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import train_test_split
import json


# **IMPORTING DATASET**

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
df = pd.read_csv("/content/drive/MyDrive/SafePath/master_clauses.csv")

# **EXPLORATORY DATA ANALYSIS**

In [4]:
df.head()

Filename  \
0  CybergyHoldingsInc_20140520_10-Q_EX-10.27_8605...   
1  EuromediaHoldingsCorp_20070215_10SB12G_EX-10.B...   
2  FulucaiProductionsLtd_20131223_10-Q_EX-10.9_83...   
3  GopageCorp_20140221_10-K_EX-10.1_8432966_EX-10...   
4  IdeanomicsInc_20160330_10-K_EX-10.26_9512211_E...   

                                    Document Name  \
0               ['MARKETING AFFILIATE AGREEMENT']   
1   ['VIDEO-ON-DEMAND CONTENT LICENSE AGREEMENT']   
2  ['CONTENT DISTRIBUTION AND LICENSE AGREEMENT']   
3           ['WEBSITE CONTENT LICENSE AGREEMENT']   
4                   ['CONTENT LICENSE AGREEMENT']   

                         Document Name-Answer  \
0               MARKETING AFFILIATE AGREEMENT   
1   VIDEO-ON-DEMAND CONTENT LICENSE AGREEMENT   
2  CONTENT DISTRIBUTION AND LICENSE AGREEMENT   
3           WEBSITE CONTENT LICENSE AGREEMENT   
4                   CONTENT LICENSE AGREEMENT   

                                             Parties  \
0  ['BIRCH FIRST GLOBAL INVESTMENTS INC.', 'MA', ...   
1  ['EuroMedia Holdings Corp.', 'Rogers', 'Rogers...   
2  ['Producer', 'Fulucai Productions Ltd.', 'Conv...   
3  ['PSiTech Corporation', 'Licensor', 'Licensee'...   
4  ['YOU ON DEMAND HOLDINGS, INC.', 'Licensor', '...   

                                      Parties-Answer  \
0  Birch First Global Investments Inc. ("Company"...   
1  Rogers Cable Communications Inc. ("Rogers"); E...   
2  CONVERGTV, INC. (“ConvergTV”); Fulucai Product...   
3  PSiTech Corporation ("Licensor"); Empirical Ve...   
4  Beijing Sun Seven Stars Culture Development Li...   

                           Agreement Date Agreement Date-Answer  \
0  ['8th day of May 2014', 'May 8, 2014']                5/8/14   
1                      ['July 11 , 2006']               7/11/06   
2                   ['November 15, 2012']              11/15/12   
3                        ['Feb 10, 2014']               2/10/14   
4                   ['December 21, 2015']              12/21/15   

                                      Effective Date Effective Date-Answer  \
0  ['This agreement shall begin upon the date of ...                   NaN   
1                                 ['July 11 , 2006']               7/11/06   
2                              ['November 15, 2012']              11/15/12   
3                                   ['Feb 10, 2014']               2/10/14   
4                              ['December 21, 2015']              12/21/15   

                                     Expiration Date  ... Liquidated Damages  \
0  ['This agreement shall begin upon the date of ...  ...                 []   
1  ['The term of this Agreement (the "Initial Ter...  ...                 []   
2                                                 []  ...                 []   
3  ['The initial term of this Agreement commences...  ...                 []   
4  ['The Term of this Agreement (the "Term") shal...  ...                 []   

  Liquidated Damages-Answer  \
0                        No   
1                        No   
2                        No   
3                        No   
4                        No   

                                   Warranty Duration Warranty Duration-Answer  \
0  ["COMPANY'S SOLE AND EXCLUSIVE LIABILITY FOR T...                      Yes   
1                                                 []                       No   
2                                                 []                       No   
3                                                 []                       No   
4                                                 []                       No   

  Insurance Insurance-Answer Covenant Not To Sue Covenant Not To Sue-Answer  \
0        []               No                  []                         No   
1        []               No                  []                         No   
2        []               No                  []                         No   
3        []               No                  []                         No 

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 510 entries, 0 to 509
Data columns (total 83 columns):
 #   Column                                      Non-Null Count  Dtype 
---  ------                                      --------------  ----- 
 0   Filename                                    510 non-null    object
 1   Document Name                               510 non-null    object
 2   Document Name-Answer                        510 non-null    object
 3   Parties                                     510 non-null    object
 4   Parties-Answer                              509 non-null    object
 5   Agreement Date                              510 non-null    object
 6   Agreement Date-Answer                       465 non-null    object
 7   Effective Date                              510 non-null    object
 8   Effective Date-Answer                       359 non-null    object
 9   Expiration Date                             510 non-null    object
 10  Expiration Date-Answer    

In [6]:
df.describe(include ="all")

Filename  \
count                                                 510   
unique                                                510   
top     KUBIENT,INC_07_02_2020-EX-10.14-MASTER SERVICE...   
freq                                                    1   

                    Document Name   Document Name-Answer  \
count                         510                    510   
unique                        274                    285   
top     ['SPONSORSHIP AGREEMENT']  SPONSORSHIP AGREEMENT   
freq                           21                     20   

                                                  Parties  \
count                                                 510   
unique                                                502   
top     ['Adviser', 'FASC', 'FEDERATED ADVISORY SERVIC...   
freq                                                    4   

                                           Parties-Answer Agreement Date  \
count                                                 509            510   
unique                                                499            462   
top     FEDERATED INVESTMENT MANAGEMENT COMPANY ("Advi...             []   
freq                                                    4             40   

       Agreement Date-Answer Effective Date Effective Date-Answer  \
count                    465            510                   359   
unique                   424            382                   328   
top               []/[]/2020             []            []/[]/2020   
freq                       5            120                     4   

       Expiration Date  ... Liquidated Damages Liquidated Damages-Answer  \
count              510  ...                510                       510   
unique             411  ...                 62                         2   
top                 []  ...                 []                        No   
freq                97  ...                449                       449   

       Warranty Duration Warranty Duration-Answer Insurance Insurance-Answer  \
count                510                      510       510              510   
unique                76                        2       167                2   
top                   []                       No        []               No   
freq                 435                      435       344              343   

       Covenant Not To Sue Covenant Not To Sue-Answer Third Party Beneficiary  \
count                  510                        510                     510   
unique                 101                          2                      33   
top                     []                         No                      []   
freq                   410                        410                     478   

       Third Party Beneficiary-Answer  
count                             510  
unique                              2  
top                                No  
freq                              477  

[4 rows x 83 columns]

In [7]:
df.shape

(510, 83)

## Each Row Represents a single contract and we have about 510 contracts (entries) and 83 columns. So conceptually there are 510 legal agreements and each agreement checked against 41 legal risk categories

In [8]:
for i, col in enumerate  (df.columns):
    print(f"{i}: {col}")

0: Filename
1: Document Name
2: Document Name-Answer
3: Parties
4: Parties-Answer
5: Agreement Date
6: Agreement Date-Answer
7: Effective Date
8: Effective Date-Answer
9: Expiration Date
10: Expiration Date-Answer
11: Renewal Term
12: Renewal Term-Answer
13: Notice Period To Terminate Renewal
14: Notice Period To Terminate Renewal- Answer
15: Governing Law
16: Governing Law-Answer
17: Most Favored Nation
18: Most Favored Nation-Answer
19: Competitive Restriction Exception
20: Competitive Restriction Exception-Answer
21: Non-Compete
22: Non-Compete-Answer
23: Exclusivity
24: Exclusivity-Answer
25: No-Solicit Of Customers
26: No-Solicit Of Customers-Answer
27: No-Solicit Of Employees
28: No-Solicit Of Employees-Answer
29: Non-Disparagement
30: Non-Disparagement-Answer
31: Termination For Convenience
32: Termination For Convenience-Answer
33: Rofr/Rofo/Rofn
34: Rofr/Rofo/Rofn-Answer
35: Change Of Control
36: Change Of Control-Answer
37: Anti-Assignment
38: Anti-Assignment-Answer
39: Reven

##To see the Column name clearly we had to print and index for each column to see the printed column names clearly.

In [9]:
df.columns.unique()

Index(['Filename', 'Document Name', 'Document Name-Answer', 'Parties',
       'Parties-Answer', 'Agreement Date', 'Agreement Date-Answer',
       'Effective Date', 'Effective Date-Answer', 'Expiration Date',
       'Expiration Date-Answer', 'Renewal Term', 'Renewal Term-Answer',
       'Notice Period To Terminate Renewal',
       'Notice Period To Terminate Renewal- Answer', 'Governing Law',
       'Governing Law-Answer', 'Most Favored Nation',
       'Most Favored Nation-Answer', 'Competitive Restriction Exception',
       'Competitive Restriction Exception-Answer', 'Non-Compete',
       'Non-Compete-Answer', 'Exclusivity', 'Exclusivity-Answer',
       'No-Solicit Of Customers', 'No-Solicit Of Customers-Answer',
       'No-Solicit Of Employees', 'No-Solicit Of Employees-Answer',
       'Non-Disparagement', 'Non-Disparagement-Answer',
       'Termination For Convenience', 'Termination For Convenience-Answer',
       'Rofr/Rofo/Rofn', 'Rofr/Rofo/Rofn-Answer', 'Change Of Control',
      

In [10]:
unique_counts = df.nunique().reset_index()
unique_counts.columns = ['Column', 'Unique values']
print(unique_counts)

                            Column  Unique values
0                         Filename            510
1                    Document Name            274
2             Document Name-Answer            285
3                          Parties            502
4                   Parties-Answer            499
..                             ...            ...
78                Insurance-Answer              2
79             Covenant Not To Sue            101
80      Covenant Not To Sue-Answer              2
81         Third Party Beneficiary             33
82  Third Party Beneficiary-Answer              2

[83 rows x 2 columns]


##We checked for unique entries in the columns, and noticed that the dataset was in a WIDE format. and It is structured like this:

| Column Type           | Meaning                |
| --------------------- | ---------------------- |
| Metadata Columns      | Contract identity info |
| Clause Columns        | Extracted clause text  |
| Clause-Answer Columns | Yes/No label           |

-`Metadata Columns` are the context that describe the contract. For example: `Filename`, `Document Name`, `Parties`, `Agreement Date`, `Effective Date`, `Expiration Date`, etc.


-`Clause Columns` (AKA Risk Indicators) are the legal risk signals. Example of these columns include: `Non-Compete`, `Exclusivity`, `Termination For Convenience`, `Uncapped Liability`, `Cap On Liability`, `Audit Rights`, `Ip Ownership Assignment`, `Liquidated Damages`, `Insurance,Third Party Beneficiary`, etc. And their corresponding Extracted clause text columns (Clause-Answer Columns) contains the actual indicators for example the preceding `Non-Compete-Answer`, `Exclusivity-Answer`, `Termination For Convenience-Answer`, and so on (if present), Empty list [] if not present. Hence `Clause Columns` is like the Risk label for the column carry indicator `Clause-Answer Columns`.



-`Clause-Answer Columns`: Like we have talked about, for every clause column, there is a matching `Clause Name - Answer`, and this is what we will primarily use for risk mapping. For example:`Non-Compete` → `Non-Compete-Answer`, `Cap On Liability → Cap On Liability-Answer, etc.  And the Values` are:`Yes` or `No`; telling us whether the clause exists in that contract or not.

## IDENTIFY CLAUSE-ANSWER COLUMNS

In [11]:
#All clause presence columns end with "-Answer"
answer_cols = [col for col in df.columns if col.endswith("-Answer")]
answer_cols

['Document Name-Answer',
 'Parties-Answer',
 'Agreement Date-Answer',
 'Effective Date-Answer',
 'Expiration Date-Answer',
 'Renewal Term-Answer',
 'Governing Law-Answer',
 'Most Favored Nation-Answer',
 'Competitive Restriction Exception-Answer',
 'Non-Compete-Answer',
 'Exclusivity-Answer',
 'No-Solicit Of Customers-Answer',
 'No-Solicit Of Employees-Answer',
 'Non-Disparagement-Answer',
 'Termination For Convenience-Answer',
 'Rofr/Rofo/Rofn-Answer',
 'Change Of Control-Answer',
 'Anti-Assignment-Answer',
 'Revenue/Profit Sharing-Answer',
 'Price Restrictions-Answer',
 'Minimum Commitment-Answer',
 'Volume Restriction-Answer',
 'Ip Ownership Assignment-Answer',
 'Joint Ip Ownership-Answer',
 'License Grant-Answer',
 'Non-Transferable License-Answer',
 'Affiliate License-Licensor-Answer',
 'Affiliate License-Licensee-Answer',
 'Unlimited/All-You-Can-Eat-License-Answer',
 'Irrevocable Or Perpetual License-Answer',
 'Source Code Escrow-Answer',
 'Post-Termination Services-Answer',
 'Au

## TOTAL NUMBER OF CLAUSE-ANSWER COLUMNS

In [12]:
print("Total Clause Types:", len(answer_cols))

Total Clause Types: 40


## FREQUENCY CALCULATION

In [13]:
clause_frequency = {}

for col in answer_cols:
    clause_name = col.replace("-Answer", "")
    yes_count = (df[col] == "Yes").sum()
    clause_frequency[clause_name] = yes_count

In [14]:
# Convert to DataFrame
freq_df = pd.DataFrame.from_dict(
    clause_frequency,
    orient="index",
    columns=["Contract_Count"]
)

In [15]:
freq_df

Contract_Count
Document Name                                   0
Parties                                         0
Agreement Date                                  0
Effective Date                                  0
Expiration Date                                 0
Renewal Term                                    0
Governing Law                                   0
Most Favored Nation                            28
Competitive Restriction Exception              76
Non-Compete                                   119
Exclusivity                                   180
No-Solicit Of Customers                        34
No-Solicit Of Employees                        59
Non-Disparagement                              38
Termination For Convenience                   183
Rofr/Rofo/Rofn                                 85
Change Of Control                             121
Anti-Assignment                               374
Revenue/Profit Sharing                        166
Price Restrictions                             15
Minimum Commitment                            165
Volume Restriction                             82
Ip Ownership Assignment                       124
Joint Ip Ownership                             46
License Grant                                 255
Non-Transferable License                      138
Affiliate License-Licensor                     23
Affiliate License-Licensee                     59
Unlimited/All-You-Can-Eat-License              17
Irrevocable Or Perpetual License               70
Source Code Escrow                             13
Post-Termination Services                     182
Audit Rights                                  214
Uncapped Liability                            111
Cap On Liability                              275
Liquidated Damages                             61
Warranty Duration                              75
Insurance                                     167
Covenant Not To Sue                           100
Third Party Beneficiary                        33

## TOP **10 MOST FREQUENT** CLAUSES BY TOTAL COUNT

In [16]:
# Sort descending
freq_df = freq_df.sort_values(
    by="Contract_Count",
    ascending=False
)

print(freq_df.head(10))

                             Contract_Count
Anti-Assignment                         374
Cap On Liability                        275
License Grant                           255
Audit Rights                            214
Termination For Convenience             183
Post-Termination Services               182
Exclusivity                             180
Insurance                               167
Revenue/Profit Sharing                  166
Minimum Commitment                      165


## TOP **10 MOST FREQUENT** CLAUSES IN %

In [17]:
total_contracts = len(df)

freq_df["Percentage"] = (
    freq_df["Contract_Count"] / total_contracts
) * 100

print(freq_df.head(10))


                             Contract_Count  Percentage
Anti-Assignment                         374   73.333333
Cap On Liability                        275   53.921569
License Grant                           255   50.000000
Audit Rights                            214   41.960784
Termination For Convenience             183   35.882353
Post-Termination Services               182   35.686275
Exclusivity                             180   35.294118
Insurance                               167   32.745098
Revenue/Profit Sharing                  166   32.549020
Minimum Commitment                      165   32.352941


# FEATURE ENGINEERING

Next we define the risk domain by grouping all legal clauses into broader conceptual risk domains.

In [18]:
risk_domain_mapping = {
    "Financial Risk": [
        "Uncapped Liability",
        "Cap On Liability",
        "Liquidated Damages",
        "Minimum Commitment",
        "Revenue/Profit Sharing",
        "Price Restrictions",
        "Most Favored Nation",
        "Volume Restriction"
    ],

    "Operational Risk": [
        "Termination For Convenience",
        "Post-Termination Services",
        "Audit Rights",
        "Anti-Assignment",
        "Change Of Control",
        "Renewal Term"
    ],

    "Strategic Restriction Risk": [
        "Non-Compete",
        "Exclusivity",
        "No-Solicit Of Employees",
        "No-Solicit Of Customers",
        "Competitive Restriction Exception",
        "Rofr/Rofo/Rofn"
    ],

    "Intellectual Property Risk": [
        "Ip Ownership Assignment",
        "Joint Ip Ownership",
        "License Grant",
        "Non-Transferable License",
        "Irrevocable Or Perpetual License",
        "Source Code Escrow",
          "Affiliate License-Licensor",
        "Affiliate License-Licensee",
        "Unlimited/All-You-Can-Eat-License"
    ],

    "Compliance / Legal Risk": [
        "Insurance",
        "Covenant Not To Sue",
        "Third Party Beneficiary",
        "Warranty Duration",
        "Non-Disparagement",
        "Governing Law"

    ]
}


CUAD clause labels are highly specific legal terms. So to make the system interpretable and useful for SMEs, clauses are grouped into higher-level risk categories such as `Financial Risk`, `Operational Risk`, `Strategic Risk`, `Intellectual Property Risk`, and `Compliance/Legal Risk`. This step represents the beginning of feature engineering, transforming raw legal categories into semantic risk features. This is essential because every clause should belong to one domain to ensure complete ontology coverage.

## DEFINING THE CLAUSE SEVERITY

In [19]:
severity_mapping = {

    # HIGH RISK (major SME exposure)
    "Uncapped Liability": "High",
    "Minimum Commitment": "High",
    "Termination For Convenience": "High",
    "Exclusivity": "High",
    "Non-Compete": "High",
    "Ip Ownership Assignment": "High",
    "Liquidated Damages": "High",

    # MEDIUM RISK (important but context-dependent)
    "Audit Rights": "Medium",
    "Revenue/Profit Sharing": "Medium",
    "Anti-Assignment": "Medium",
    "Cap On Liability": "Medium",
    "Volume Restriction": "Medium",
    "Most Favored Nation": "Medium",
    "Rofr/Rofo/Rofn": "Medium",
    "Affiliate License-Licensor": "Medium",
    "Affiliate License-Licensee": "Medium",
    "Change Of Control": "Medium",
    "No-Solicit Of Employees": "Medium",
    "No-Solicit Of Customers": "Medium",
    "Competitive Restriction Exception": "Medium",
    "Post-Termination Services": "Medium",
    "Price Restrictions": "Medium",
    "Irrevocable Or Perpetual License": "Medium",
    "Source Code Escrow": "Medium",

    # LOW RISK (legal structure / informational)
    "Warranty Duration": "Low",
    "Non-Disparagement": "Low",
    "Unlimited/All-You-Can-Eat-License": "Low",
    "Renewal Term": "Low",
    "Notice Period To Terminate Renewal": "Low",
    "Governing Law": "Low",
    "Joint Ip Ownership": "Low",
    "License Grant": "Low",
    "Non-Transferable License": "Low",
    "Insurance": "Low",
    "Covenant Not To Sue": "Low",
    "Third Party Beneficiary": "Low"
}


We assigned an impact level (High/Medium/Low) to each clause because not all clauses pose equal risk to SMEs. Severity reflects potential financial, operational, legal, or strategic impact if the clause is unfavorable. This allows the system to prioritize warnings for users.

In [20]:
severity_score_map = {
    "High": 3,
    "Medium": 2,
    "Low": 1
}


The Severity Score map converts qualitative severity labels into numeric values. This is important because numeric scores enable quantitative analysis such as:

    Risk Score = Frequency × Severity

This allows ranking clauses by overall risk importance,mapping High to 3,  Medium to 2 and Low to 1


## To add Severity to Frequency Table

In [21]:
freq_df["Severity"] = freq_df.index.map(
    lambda x: severity_mapping.get(x, "Low")
)

The Purpose of adding severity to frequency table is to assigns a qualitative risk level (High/Medium/Low) to each clause using the predefined severity mapping.
To ensure that we use the map() to look up each clause in severity_mapping and use the get(x, "Low") to  ensure missing clauses default to Low.




In [22]:
freq_df

Contract_Count  Percentage Severity
Anti-Assignment                               374   73.333333   Medium
Cap On Liability                              275   53.921569   Medium
License Grant                                 255   50.000000      Low
Audit Rights                                  214   41.960784   Medium
Termination For Convenience                   183   35.882353     High
Post-Termination Services                     182   35.686275   Medium
Exclusivity                                   180   35.294118     High
Insurance                                     167   32.745098      Low
Revenue/Profit Sharing                        166   32.549020   Medium
Minimum Commitment                            165   32.352941     High
Non-Transferable License                      138   27.058824      Low
Ip Ownership Assignment                       124   24.313725     High
Change Of Control                             121   23.725490   Medium
Non-Compete                                   119   23.333333     High
Uncapped Liability                            111   21.764706     High
Covenant Not To Sue                           100   19.607843      Low
Rofr/Rofo/Rofn                                 85   16.666667   Medium
Volume Restriction                             82   16.078431   Medium
Competitive Restriction Exception              76   14.901961   Medium
Warranty Duration                              75   14.705882      Low
Irrevocable Or Perpetual License               70   13.725490   Medium
Liquidated Damages                             61   11.960784     High
No-Solicit Of Employees                        59   11.568627   Medium
Affiliate License-Licensee                     59   11.568627   Medium
Joint Ip Ownership                             46    9.019608      Low
Non-Disparagement                              38    7.450980      Low
No-Solicit Of Customers                        34    6.666667   Medium
Third Party Beneficiary                        33    6.470588      Low
Most Favored Nation                            28    5.490196   Medium
Affiliate License-Licensor                     23    4.509804   Medium
Unlimited/All-You-Can-Eat-License              17    3.333333      Low
Price Restrictions                             15    2.941176   Medium
Source Code Escrow                             13    2.549020   Medium
Governing Law                                   0    0.000000      Low
Expiration Date                                 0    0.000000      Low
Renewal Term                                    0    0.000000      Low
Parties                                         0    0.000000      Low
Agreement Date                                  0    0.000000      Low
Document Name                                   0    0.000000      Low
Effective Date                                  0    0.000000      Low

##To add Numeric Scores

In [23]:
freq_df["Severity_Score"] = freq_df["Severity"].map(
    severity_score_map
)

This converts qualitative severity labels into numbers because mathematical calculations require numeric values and we need  Severity score to compute the Risk Score

In [24]:
freq_df

Contract_Count  Percentage Severity  \
Anti-Assignment                               374   73.333333   Medium   
Cap On Liability                              275   53.921569   Medium   
License Grant                                 255   50.000000      Low   
Audit Rights                                  214   41.960784   Medium   
Termination For Convenience                   183   35.882353     High   
Post-Termination Services                     182   35.686275   Medium   
Exclusivity                                   180   35.294118     High   
Insurance                                     167   32.745098      Low   
Revenue/Profit Sharing                        166   32.549020   Medium   
Minimum Commitment                            165   32.352941     High   
Non-Transferable License                      138   27.058824      Low   
Ip Ownership Assignment                       124   24.313725     High   
Change Of Control                             121   23.725490   Medium   
Non-Compete                                   119   23.333333     High   
Uncapped Liability                            111   21.764706     High   
Covenant Not To Sue                           100   19.607843      Low   
Rofr/Rofo/Rofn                                 85   16.666667   Medium   
Volume Restriction                             82   16.078431   Medium   
Competitive Restriction Exception              76   14.901961   Medium   
Warranty Duration                              75   14.705882      Low   
Irrevocable Or Perpetual License               70   13.725490   Medium   
Liquidated Damages                             61   11.960784     High   
No-Solicit Of Employees                        59   11.568627   Medium   
Affiliate License-Licensee                     59   11.568627   Medium   
Joint Ip Ownership                             46    9.019608      Low   
Non-Disparagement                              38    7.450980      Low   
No-Solicit Of Customers                        34    6.666667   Medium   
Third Party Beneficiary                        33    6.470588      Low   
Most Favored Nation                            28    5.490196   Medium   
Affiliate License-Licensor                     23    4.509804   Medium   
Unlimited/All-You-Can-Eat-License              17    3.333333      Low   
Price Restrictions                             15    2.941176   Medium   
Source Code Escrow                             13    2.549020   Medium   
Governing Law                                   0    0.000000      Low   
Expiration Date                                 0    0.000000      Low   
Renewal Term                                    0    0.000000      Low   
Parties                                         0    0.000000      Low   
Agreement Date                                  0    0.000000      Low   
Document Name                                   0    0.000000      Low   
Effective Date                                  0    0.000000      Low   

                                   Severity_Score  
Anti-Assignment                                 2  
Cap On Liability                                2  
License Grant                                   1  
Audit Rights                                    2  
Termination For Convenience                     3  
Post-Termination Services                       2  
Exclusivity                                     3  
Insurance                                       1  
Revenue/Profit Sharing                          2  
Minimum Commitment                              3  
Non-Transferable License                        1  
Ip Ownership Assignment                         3  
Change Of Control                               2  
Non-Compete                                     3  
Uncapped Liability                              3  
Covenant Not To Sue                             1  
Rofr/Rofo/Rofn                                  2  
Volume Restriction                              2  
Competitive 

##Calculate Risk Score

In [25]:
freq_df["Risk_Score"] = (
    freq_df["Contract_Count"] *
    freq_df["Severity_Score"]
)


## Risk Score Formula:
     Risk Score = Contract_Count × Severity_Score

This combine clause prevalence and severity into one quantitative risk metric.
And Higher Score equals to clause that is both common and impactful.


In [26]:
freq_df

Contract_Count  Percentage Severity  \
Anti-Assignment                               374   73.333333   Medium   
Cap On Liability                              275   53.921569   Medium   
License Grant                                 255   50.000000      Low   
Audit Rights                                  214   41.960784   Medium   
Termination For Convenience                   183   35.882353     High   
Post-Termination Services                     182   35.686275   Medium   
Exclusivity                                   180   35.294118     High   
Insurance                                     167   32.745098      Low   
Revenue/Profit Sharing                        166   32.549020   Medium   
Minimum Commitment                            165   32.352941     High   
Non-Transferable License                      138   27.058824      Low   
Ip Ownership Assignment                       124   24.313725     High   
Change Of Control                             121   23.725490   Medium   
Non-Compete                                   119   23.333333     High   
Uncapped Liability                            111   21.764706     High   
Covenant Not To Sue                           100   19.607843      Low   
Rofr/Rofo/Rofn                                 85   16.666667   Medium   
Volume Restriction                             82   16.078431   Medium   
Competitive Restriction Exception              76   14.901961   Medium   
Warranty Duration                              75   14.705882      Low   
Irrevocable Or Perpetual License               70   13.725490   Medium   
Liquidated Damages                             61   11.960784     High   
No-Solicit Of Employees                        59   11.568627   Medium   
Affiliate License-Licensee                     59   11.568627   Medium   
Joint Ip Ownership                             46    9.019608      Low   
Non-Disparagement                              38    7.450980      Low   
No-Solicit Of Customers                        34    6.666667   Medium   
Third Party Beneficiary                        33    6.470588      Low   
Most Favored Nation                            28    5.490196   Medium   
Affiliate License-Licensor                     23    4.509804   Medium   
Unlimited/All-You-Can-Eat-License              17    3.333333      Low   
Price Restrictions                             15    2.941176   Medium   
Source Code Escrow                             13    2.549020   Medium   
Governing Law                                   0    0.000000      Low   
Expiration Date                                 0    0.000000      Low   
Renewal Term                                    0    0.000000      Low   
Parties                                         0    0.000000      Low   
Agreement Date                                  0    0.000000      Low   
Document Name                                   0    0.000000      Low   
Effective Date                                  0    0.000000      Low   

                                   Severity_Score  Risk_Score  
Anti-Assignment                                 2         748  
Cap On Liability                                2         550  
License Grant                                   1         255  
Audit Rights                                    2         428  
Termination For Convenience                     3         549  
Post-Termination Services                       2         364  
Exclusivity                                     3         540  
Insurance                                       1         167  
Revenue/Profit Sharing                          2         332  
Minimum Commitment                              3         495  
Non-Transferable License                        1         138  
Ip Ownership Assignment                         3         372  
Change Of Control                               2         242  
Non-Compete                                     3         357  
Uncapped Liability                      

## To generate Risk-Weighted Ranking

In [27]:
risk_ranked_df = freq_df.sort_values(
    by="Risk_Score",
    ascending=False
)

Ranked clauses from highest to lowest risk priority by sorting dataframe by Risk_Score descending.

In [28]:
risk_ranked_df.head(15)


Contract_Count  Percentage Severity  \
Anti-Assignment                         374   73.333333   Medium   
Cap On Liability                        275   53.921569   Medium   
Termination For Convenience             183   35.882353     High   
Exclusivity                             180   35.294118     High   
Minimum Commitment                      165   32.352941     High   
Audit Rights                            214   41.960784   Medium   
Ip Ownership Assignment                 124   24.313725     High   
Post-Termination Services               182   35.686275   Medium   
Non-Compete                             119   23.333333     High   
Uncapped Liability                      111   21.764706     High   
Revenue/Profit Sharing                  166   32.549020   Medium   
License Grant                           255   50.000000      Low   
Change Of Control                       121   23.725490   Medium   
Liquidated Damages                       61   11.960784     High   
Rofr/Rofo/Rofn                           85   16.666667   Medium   

                             Severity_Score  Risk_Score  
Anti-Assignment                           2         748  
Cap On Liability                          2         550  
Termination For Convenience               3         549  
Exclusivity                               3         540  
Minimum Commitment                        3         495  
Audit Rights                              2         428  
Ip Ownership Assignment                   3         372  
Post-Termination Services                 2         364  
Non-Compete                               3         357  
Uncapped Liability                        3         333  
Revenue/Profit Sharing                    2         332  
License Grant                             1         255  
Change Of Control                         2         242  
Liquidated Damages                        3         183  
Rofr/Rofo/Rofn                            2         170

##Creating a Train / Test Split set

In [29]:
train_df, test_df = train_test_split(
    df,
    test_size=0.3,
    random_state=42,
    shuffle=True,
    stratify=df["Termination For Convenience-Answer"]
)

print("Train size:", train_df.shape)
print("Test size:", test_df.shape)

Train size: (357, 83)
Test size: (153, 83)


#TASK 2: SYSTEM EXECUTION

##Data Cleaning

In [30]:
def clean_text(text):
    text = str(text)
    text = re.sub(r"[\[\]']", "", text)
    return text


We define this function to clean  Clause Text since most of our clause text samples are list-like strings.

##Clause Text Exploration

In [31]:
clause = "Termination For Convenience"

examples = train_df[
    train_df[f"{clause}-Answer"] == "Yes"
][clause]

print(examples.head(10))


398    ["Upon the occurrence of any of the following,...
492    ['FUSION may terminate this Agreement by provi...
204    ['The Company reserves the right in its sole a...
325    ['At any time during a renewed term, this Agre...
464    ['Owner may terminate this Agreement at any ti...
272    ["With respect to Beacon Optofluidic Machines ...
178    ["Pretzel  Time  reserves  the right to  suspe...
358              ['Termination for Convenience. [***].']
236    ["Except as may be provided in any Schedule, M...
375    ['The Parties may terminate its performance of...
Name: Termination For Convenience, dtype: object


We started with ONE clause first, `Termination For Convenience`clause for Clause Text Exploration. Extract real corpus examples from the clause column and printed about 10 samples to see how they look like.

##Extract Lemmas Using SPACY (Rule Discovery)

In [32]:
texts = [clean_text(t) for t in examples]

lemmas = []

for text in texts:
    doc = nlp(text)

    for token in doc:
        if token.is_stop or token.is_punct:
            continue
        lemmas.append(token.lemma_.lower())


print(Counter(lemmas).most_common(20))

[('agreement', 183), ('notice', 168), ('terminate', 164), ('write', 139), ('day', 98), ('party', 92), (' ', 90), ('prior', 88), ('time', 79), ('shall', 60), ('termination', 56), ('provide', 53), ('date', 42), ('right', 41), ('term', 38), ('month', 37), ('give', 36), ('30', 34), ('reason', 29), ('company', 27)]


 We had to use spaCy to normalize legal language through lemmatization, identify common verbs and noun phrases and very importantly reveal recurring linguistic patterns in clauses. This is so important because these patterns would guide the regex rule design towards a deterministic rule-based detection approach.

##Extract Key Noun Phrases

In [33]:
noun_phrases = []

for text in texts:
    doc = nlp(text)

    for chunk in doc.noun_chunks:
        noun_phrases.append(chunk.text.lower())

print(Counter(noun_phrases).most_common(20))


[('this agreement', 152), ('any time', 72), ('notice', 43), ('either party', 35), ('written notice', 28), ('the other party', 28), ('prior written notice', 25), ('cause', 24), ('the right', 21), ('any reason', 20), ('termination', 17), ('the effective date', 17), ('the company', 17), ('any', 16), ('section', 12), ('convenience', 11), ('thirty (30) days', 11), ('the agreement', 11), ('such termination', 11), ('*', 11)]


In [34]:
# PHRASE FREQUENCY
phrase_freq = Counter(noun_phrases)

# REGEX GENERATION
TOP_N = 10

top_phrases = [
    p for p, c in phrase_freq.most_common(TOP_N)
]

ignore_phrases = [
    "this agreement",
    "the agreement",
    "section",
    "which"
]


top_phrases = [
    p for p, c in phrase_freq.most_common(30)
    if p not in ignore_phrases
][:10]


print(top_phrases)

['any time', 'notice', 'either party', 'written notice', 'the other party', 'prior written notice', 'cause', 'the right', 'any reason', 'termination']


##Extract Verb + Nearby Phrase

In [35]:
verb_phrase_pairs = []

for text in texts:
    doc = nlp(text)

    for token in doc:

        # focus on verbs
        if token.pos_ == "VERB":

            # collect nearby noun chunks
            for chunk in doc.noun_chunks:

                # distance filter (same sentence area)
                if abs(token.i - chunk.root.i) < 6:

                    pair = (
                        token.lemma_.lower(),
                        chunk.text.lower()
                    )

                    verb_phrase_pairs.append(pair)


Legal clauses are often defined by actions (verbs) combined with contextual phrases. This step uses spaCy to identify verbs and nearby noun phrases to automatically suggest regex pattern structures. These patterns improve detection precision compared to standalone keywords.

##VERB–PHRASE FREQUENCY ANALYSIS

In [36]:
pair_freq = Counter(verb_phrase_pairs)

print(pair_freq.most_common(20))


[(('terminate', 'this agreement'), 117), (('write', 'notice'), 35), (('terminate', 'either party'), 29), (('write', 'written notice'), 27), (('terminate', 'any time'), 26), (('write', 'prior written notice'), 25), (('write', 'the other party'), 21), (('terminate', 'the right'), 16), (('have', 'the right'), 16), (('write', 'any time'), 11), (('write', 'this agreement'), 9), (('terminate', 'cause'), 9), (('give', 'any time'), 9), (('write', 'sixty (60) days'), 9), (('write', 'thirty (30) days'), 8), (('write', 'ninety (90) days'), 8), (('give', 'this agreement'), 8), (('give', 'written notice'), 7), (('terminate', 'the agreement'), 7), (('write', 'termination'), 6)]


This output a  ranked list of the most common verb–phrase pairs, which serves as input for regex pattern generation. It counts how frequently specific verb–phrase combinations appears across clause examples. Legal clauses often use recurring action patterns (e.g., "terminate + written notice") so a high-frequency pair would represent strong linguistic signals for clause identification.


##Turn Into Regex Suggestions

In [37]:
regex_suggestions = []

for (verb, phrase), count in pair_freq.most_common(10):

    phrase_regex = ".*".join(phrase.split())

    pattern = f"{verb}.*{phrase_regex}"

    regex_suggestions.append(pattern)

print(regex_suggestions)

['terminate.*this.*agreement', 'write.*notice', 'terminate.*either.*party', 'write.*written.*notice', 'terminate.*any.*time', 'write.*prior.*written.*notice', 'write.*the.*other.*party', 'terminate.*the.*right', 'have.*the.*right', 'write.*any.*time']


This output candidate regex patterns used for clause detection by automatically convert frequent verb–phrase patterns into candidate regex rules.

 And this is acheived by:
1. Select the most common verb–phrase pairs.
2. Replace spaces with '.*' to allow flexible wording.
3. Combine verb and phrase into regex structure.


##Automated Rule Dictionary Generation

In [38]:
regex_rules = {
    "Termination For Convenience": regex_suggestions
}

Instead of manually defining regex patterns for each clause, the rules can be automatically constructed from the most frequent verb–phrase patterns extracted using spaCy. This allows the detection logic to remain data-driven and reproducible across different clause categories. The automatically generated regex suggestions are grouped under their respective clause names to form the final rule dictionary used by the detection engine. This approach reduces manual bias, improves scalability for multiple clauses, and ensures that rule patterns reflect real contractual language observed in the dataset. While automation accelerates rule development, generated patterns are reviewed to remove overly generic or noisy expressions before deployment.

##Rule-Based Clause Detection Function

In [39]:
def detect_clause(text, patterns):

    text = str(text).lower()

    for pattern in patterns:
        if re.search(pattern, text):
            return 1   # detected

    return 0


This function performs rule-based clause detection using regex patterns. Each contract text is normalized to lowercase to ensure consistent matching. The function checks whether any regex pattern associated with the target clause appears in the text. If a match is found, the function returns 1 (clause detected); otherwise, it returns 0. This binary output enables comparison with ground-truth labels during evaluation.

##Generate Rule-Based Predictions and Ground Truth Labels

In [40]:
patterns = regex_rules[clause]

eval_df = test_df.copy()

# predicted by rules
eval_df["Predicted"] = eval_df[clause].apply(
    lambda x: detect_clause(x, patterns)
)

# actual CUAD labels
eval_df["Actual"] = eval_df[f"{clause}-Answer"].apply(
    lambda x: 1 if x == "Yes" else 0
)


This step applies the detection function across all clause texts to produce rule-based predictions. The predicted values are then compared against CUAD’s labeled outputs, which serve as the ground truth. Both predictions and actual labels are represented as binary values (1 = clause present, 0 = clause absent), enabling quantitative evaluation of rule performance.

#TASK 3: Evaluation & Metrics


##Evaluating The Detection Performance (Precision and Recall)

In [41]:
precision = precision_score(
    eval_df["Actual"],
    eval_df["Predicted"],
    zero_division=0
)

recall = recall_score(
    eval_df["Actual"],
    eval_df["Predicted"]
)

print("Precision:", precision)
print("Recall:", recall)

Precision: 1.0
Recall: 0.9090909090909091


Precision and recall are used to measure the effectiveness of the rule-based detection system. Precision indicates how many detected clauses are truly correct, while recall measures how many true clauses are successfully captured by the rules. In the `Termination For Convenience` clause, we got a precision of 1.0, indicating no false positives, while recall was 0.868 suggesting that most, but not all, the true clause occurrences were detected.

##False Positive Analysis

In [42]:
false_positives = eval_df[
    (eval_df["Predicted"] == 1) &
    (eval_df["Actual"] == 0)
]

print(false_positives[[clause]].head())

Empty DataFrame
Columns: [Termination For Convenience]
Index: []


False positives occur when rules detect a clause where none exists according to CUAD labels. Inspecting false positives helps identify overly broad regex patterns.

In this case, the result is an empty dataframe, indicating that the current rule set produced no false positive detections for this clause.

##Coverage Testing for the Rule-Based Detection

In [43]:
coverage = eval_df["Predicted"].mean()
print("Coverage:", coverage)

Coverage: 0.32679738562091504


Coverage measures how frequently the rules trigger across the dataset. It provides insight into how broadly a clause appears or how sensitive the detection rules are. A coverage score of approximately 31% indicates that the rules identified the `Termination For Convenience` clause in nearly one-third of contracts.

##Confusion Matrix for Rule Performance

In [44]:
cm = confusion_matrix(
    eval_df["Actual"],
    eval_df["Predicted"]
)
print(cm)

[[98  0]
 [ 5 50]]


The confusion matrix summarizes rule performance by showing true negatives, false positives, false negatives, and true positives. This provides a complete view of detection outcomes. In this case, the matrix shows zero false positives and some false negatives, suggesting the rules are highly precise but may miss on  `Termination For Convenience` clause variations.

#Task 4: Multi-Clause Automation

After validating the rule-generation process on a single clause, this workflow is intended to generalized into reusable functions to enable scalable rule creation across multiple clauses. Each clause would undergo the same pipeline: example extraction, linguistic pattern discovery using spaCy, regex suggestion generation, and rule evaluation. This ensures methodological consistency across clauses and supports reproducible rule-based detection at scale.

## Defining The Target Clauses

In [45]:
target_clauses = [
    "Anti-Assignment",
    "Cap On Liability",
    "Termination For Convenience",
    "Exclusivity",
    "Minimum Commitment",
    "Audit Rights",
    "Ip Ownership Assignment",
    "Post-Termination Services",
    "Non-Compete",
    "Uncapped Liability",
    "Revenue/Profit Sharing",
    "License Grant",
    "Change Of Control",
    "Liquidated Damages",
    "Rofr/Rofo/Rofn"
]


The target clauses represent the highest-priority risks selected from the risk-weighted ranking generated in Task 1. These clauses combine high frequency and significant legal impact for SMEs and startups, making them suitable for the first version of the rule-based detection system. Limiting the scope to a defined set of clauses ensures manageable development, consistent evaluation, and focused rule engineering before scaling to all clause categories.

##Extract Examples

In [46]:
def get_clause_examples(train_df, clause):
    return train_df[train_df[f"{clause}-Answer"] == "Yes"][clause]


This function extracts clause text samples from the dataset where the clause is labeled as present (“Yes”). These examples serve as the real corpus used for linguistic pattern discovery. By relying only on labeled positive examples, the rule-generation process remains grounded in authentic legal language rather than manually constructed assumptions.

##Generate Verb–Phrase Frequencies

In [47]:
def get_verb_phrase_freq(texts, nlp):

    verb_phrase_pairs = []

    for text in texts:

        doc = nlp(str(text))

        for token in doc:
            if token.pos_ == "VERB":

                for chunk in doc.noun_chunks:
                    if abs(token.i - chunk.root.i) < 6:

                        pair = (
                            token.lemma_.lower(),
                            chunk.text.lower()
                        )

                        verb_phrase_pairs.append(pair)

    return Counter(verb_phrase_pairs)


This step identifies recurring verb–phrase relationships within clause examples using spaCy’s linguistic parsing capabilities. Legal clauses are often structured around actions (verbs) combined with contextual phrases, such as “terminate + written notice.” Counting these patterns reveals high-frequency linguistic structures that serve as strong indicators for clause detection and later regex generation.

##Generate Regex Suggestions

In [48]:
def generate_regex_suggestions(pair_freq, top_n=10):

    regex_suggestions = []

    for (verb, phrase), count in pair_freq.most_common(top_n):
        # Escape special regex characters in the phrase and verb
        escaped_phrase_parts = [re.escape(part) for part in phrase.split()]
        phrase_regex = ".*".join(escaped_phrase_parts)

        escaped_verb = re.escape(verb)
        pattern = f"{escaped_verb}.*{phrase_regex}"

        regex_suggestions.append(pattern)

    return regex_suggestions

Frequent verb–phrase patterns are automatically transformed into candidate regex rules. Spaces within phrases are replaced with flexible wildcard matching (.*) to allow variation in legal wording. Regex escaping is applied to ensure that special characters are interpreted safely. These automatically generated patterns form the initial rule candidates, which are later reviewed and refined for reliability.

##Automate For All Clauses

In [49]:
all_regex_rules = {}

for clause in target_clauses:

    print(f"Processing: {clause}")

    # extract examples
    examples = get_clause_examples(train_df, clause)

    texts = [clean_text(t) for t in examples]

    # verb–phrase discovery
    pair_freq = get_verb_phrase_freq(texts, nlp)

    # regex generation
    regex_suggestions = generate_regex_suggestions(pair_freq)

    # store rules
    all_regex_rules[clause] = regex_suggestions

Processing: Anti-Assignment
Processing: Cap On Liability
Processing: Termination For Convenience
Processing: Exclusivity
Processing: Minimum Commitment
Processing: Audit Rights
Processing: Ip Ownership Assignment
Processing: Post-Termination Services
Processing: Non-Compete
Processing: Uncapped Liability
Processing: Revenue/Profit Sharing
Processing: License Grant
Processing: Change Of Control
Processing: Liquidated Damages
Processing: Rofr/Rofo/Rofn


The rule-generation workflow is automated across all target clauses to ensure consistency and scalability. For each clause, the pipeline performs example extraction, linguistic pattern discovery, and regex suggestion generation. The resulting patterns are stored in a centralized rule dictionary, enabling efficient reuse during detection and evaluation while minimizing manual rule creation.

In [50]:
all_regex_rules['Rofr/Rofo/Rofn']

['write.*written.*notice',
 'have.*the.*right',
 'provide.*written.*notice',
 'enter.*an.*agreement',
 'negotiate.*good.*faith',
 'exercise.*its.*option',
 'include.*which',
 'provide.*bii',
 'write.*it',
 'set.*section']

In [51]:
all_regex_rules['Cap On Liability']

['advise.*the.*possibility',
 'relate.*this.*agreement',
 'include.*negligence',
 'arise.*this.*agreement',
 'include.*contract',
 'arise.*connection',
 'lose.*lost.*profits',
 'punitive.*damages',
 'give.*rise',
 'include.*damages']

In [52]:
all_regex_rules['Exclusivity']

['have.*the.*exclusive.*right',
 'sell.*market',
 'offer.*sale',
 'set.*this.*agreement',
 'sublicense.*the.*right',
 'include.*limitation',
 'grant.*the.*right',
 'grant.*sublicenses',
 'set.*the.*terms',
 'sell.*the.*exclusive.*right']

In [53]:
def evaluate_clause(test_df, clause, patterns):

    eval_df = test_df.copy()

    eval_df["Predicted"] = eval_df[clause].apply(
        lambda x: detect_clause(x, patterns)
    )

    eval_df["Actual"] = eval_df[f"{clause}-Answer"].apply(
        lambda x: 1 if x == "Yes" else 0
    )

    precision = precision_score(
        eval_df["Actual"],
        eval_df["Predicted"],
        zero_division=0
    )

    recall = recall_score(
        eval_df["Actual"],
        eval_df["Predicted"]
    )

    coverage = eval_df["Predicted"].mean()

    return precision, recall, coverage


This function evaluates rule performance by comparing rule-based predictions against CUAD’s labeled ground truth. Precision measures the accuracy of detected clauses, recall measures the proportion of true clauses successfully identified, and coverage indicates how frequently rules activate across the dataset. These metrics provide quantitative evidence of rule effectiveness and guide future refinement.

##Run Evaluation for ALL Clauses

In [54]:
results = []

for clause in target_clauses:

    patterns = all_regex_rules[clause]

    precision, recall, coverage = evaluate_clause(
        test_df,
        clause,
        patterns
    )

    results.append({
        "Clause": clause,
        "Precision": precision,
        "Recall": recall,
        "Coverage": coverage
    })

results_df = pd.DataFrame(results)

print(results_df)


                         Clause  Precision    Recall  Coverage
0               Anti-Assignment        1.0  0.885714  0.607843
1              Cap On Liability        1.0  0.640000  0.313725
2   Termination For Convenience        1.0  0.909091  0.326797
3                   Exclusivity        1.0  0.482759  0.183007
4            Minimum Commitment        1.0  0.301887  0.104575
5                  Audit Rights        1.0  0.706897  0.267974
6       Ip Ownership Assignment        1.0  0.823529  0.183007
7     Post-Termination Services        1.0  0.490196  0.163399
8                   Non-Compete        1.0  0.470588  0.104575
9            Uncapped Liability        1.0  0.666667  0.130719
10       Revenue/Profit Sharing        1.0  0.388889  0.137255
11                License Grant        1.0  0.828947  0.411765
12            Change Of Control        1.0  0.571429  0.130719
13           Liquidated Damages        1.0  0.578947  0.071895
14               Rofr/Rofo/Rofn        1.0  0.640000  0

Evaluation is executed across all selected clauses to provide a comparative performance overview. Running the same evaluation pipeline for multiple clauses ensures consistency and reveals which clauses have strong or weak rule coverage. The resulting table supports systematic improvement of regex rules and helps prioritize refinement efforts.

#Task 5: MODEL PACKAGING

This stage transitions the project from research experimentation into a deployable decision-support engine. Instead of keeping rules and logic inside exploratory notebooks, the model components are exported into reusable artifacts that other development teams can integrate into applications.

##Exporting The Regex Rules

In [64]:
save_path = "/content/drive/MyDrive/SafePath"


with open(f"{save_path}/rules.json", "w") as f:
    json.dump(all_regex_rules, f, indent=2)

The generated regex rules are exported into a JSON file to allow external systems (e.g., mobile apps or backend APIs) to load the detection logic without relying on notebook execution. This separates rule logic from analysis code and supports scalable deployment.

##Exporting The Severity Mapping

In [65]:
with open(f"{save_path}/severity_mapping.json", "w") as f:
    json.dump(severity_mapping, f, indent=2)

Severity mappings are exported as a standalone resource so that risk prioritization can be applied consistently across platforms. This enables the user interface to display meaningful warnings based on predefined risk levels.

##Create Clause Into Domain Lookup

In [57]:
# Build clause → domain lookup
clause_to_domain = {}

for domain, clauses in risk_domain_mapping.items():
    for c in clauses:
        clause_to_domain[c] = domain


The risk taxonomy is reversed to create a clause-to-domain lookup dictionary. This transformation allows the system to quickly identify the broader risk category associated with each detected clause during runtime.

In [58]:
clause_to_domain

{'Uncapped Liability': 'Financial Risk',
 'Cap On Liability': 'Financial Risk',
 'Liquidated Damages': 'Financial Risk',
 'Minimum Commitment': 'Financial Risk',
 'Revenue/Profit Sharing': 'Financial Risk',
 'Price Restrictions': 'Financial Risk',
 'Most Favored Nation': 'Financial Risk',
 'Volume Restriction': 'Financial Risk',
 'Termination For Convenience': 'Operational Risk',
 'Post-Termination Services': 'Operational Risk',
 'Audit Rights': 'Operational Risk',
 'Anti-Assignment': 'Operational Risk',
 'Change Of Control': 'Operational Risk',
 'Renewal Term': 'Operational Risk',
 'Non-Compete': 'Strategic Restriction Risk',
 'Exclusivity': 'Strategic Restriction Risk',
 'No-Solicit Of Employees': 'Strategic Restriction Risk',
 'No-Solicit Of Customers': 'Strategic Restriction Risk',
 'Competitive Restriction Exception': 'Strategic Restriction Risk',
 'Rofr/Rofo/Rofn': 'Strategic Restriction Risk',
 'Ip Ownership Assignment': 'Intellectual Property Risk',
 'Joint Ip Ownership': 'Inte

##Create Severity Templates

In [59]:
severity_templates = {
    "High": "This clause may significantly affect your rights or obligations.",
    "Medium": "This clause may require attention depending on your situation.",
    "Low": "This clause is generally informational but still important to review."
}


Severity templates define standardized explanation styles for each risk level. These templates ensure that generated user warnings are consistent, interpretable, and aligned with the system’s transparency objectives.

##Auto-generate Messages

In [60]:
warning_messages = {}

for clause in target_clauses:

    severity = severity_mapping.get(clause, "Low")
    domain = clause_to_domain.get(clause, "General Risk")

    warning_messages[clause] = (
    f"{severity_templates[severity]} "
    f"This relates to {domain.lower()}."
)


User-facing warning messages are generated automatically using clause severity and risk domain information. This avoids manual hardcoding and ensures scalability as additional clauses or risk categories are introduced.

##Exporting The Warning  Messages JSON file

In [67]:
with open(f"{save_path}/warning_messages.json", "w") as f:
    json.dump(warning_messages, f, indent=2)

##Analyze Agreement Function

In [62]:
def traffic_light(severity):
    if severity == "High":
        return "red"
    elif severity == "Medium":
        return "yellow"
    else:
        return "green"

def analyze_agreement(text, rules, severity_mapping, warning_messages):

    text = str(text).lower()
    flags = []

    for clause, patterns in rules.items():

        for pattern in patterns:

            if re.search(pattern, text):

                severity = severity_mapping.get(clause, "Low")

                flags.append({
                    "clause": clause,
                    "severity": severity,
                    "traffic_light": traffic_light(severity),
                    "message": warning_messages.get(clause, "")
                })

                break
 # --- Overall Risk Summary ---
    high_count = sum(1 for f in flags if f["severity"] == "High")
    medium_count = sum(1 for f in flags if f["severity"] == "Medium")

    if high_count > 0:
        overall_risk = "High"
        summary = f"{high_count} high-risk clause(s) detected."
    elif medium_count > 0:
        overall_risk = "Medium"
        summary = f"{medium_count} medium-risk clause(s) detected."
    else:
        overall_risk = "Low"
        summary = "No major risk clauses detected."

    return {
        "flags": flags,
        "overall_risk": overall_risk,
        "summary": summary
    }

The analyze_agreement() function represents the core decision engine of the system. It processes agreement text, applies rule-based clause detection, assigns severity and visual risk indicators, and generates user-facing explanations. The function outputs structured JSON-ready results that can be directly consumed by mobile or web applications.

And also the overall risk summary is computed by aggregating detected clause severities. This provides users with a quick high-level assessment before reviewing individual warnings, reducing cognitive load and supporting faster decision-making.